pkg/controller/utils/imageset.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operator "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/components"
+	"github.com/tigera/operator/pkg/render"
+)
+
+// ApplyImageSet gets the appropriate ImageSet and applies it to
+// the components passed in.
+func ApplyImageSet(ctx context.Context, c client.Client, v operator.ProductVariant, comps ...render.Component) error {
+
+	imageSet, err := getImageSet(ctx, c, v)
+	if err != nil {
+		return err
+	}
+
+	if err = validateImageSet(imageSet); err != nil {
+		return err
+	}
+
+	errMsgs := []string{}
+	for _, component := range comps {
+		err = component.ResolveImages(imageSet)
+		if err != nil {
+			errMsgs = append(errMsgs, err.Error())
+		}
+	}
+
+	if len(errMsgs) == 0 {
+		return nil
+	}
+
+	return fmt.Errorf("Invalid ImageSet: %s", strings.Join(errMsgs, ", "))
+}
+
+func AddImageSetWatch(c controller.Controller) error {
+	return c.Watch(&source.Kind{Type: &operator.ImageSet{}}, &handler.EnqueueRequestForObject{})
+}
+
+func getSetName(v operator.ProductVariant) string {
+	setName := fmt.Sprintf("calico-%s", components.CalicoRelease)
+	if v == operator.TigeraSecureEnterprise {
+		setName = fmt.Sprintf("enterprise-%s", components.EnterpriseRelease)
+	}
+	return setName
+}
+
+// getImageSet finds the ImageSet CR for specified variant and for the correct .
+func getImageSet(ctx context.Context, cli client.Client, v operator.ProductVariant) (*operator.ImageSet, error) {
+	isl := &operator.ImageSetList{}
+
+	// List the ImageSets because if any exist then we will require the
+	// existence of one for the expected version of the operator running
+	// and the variant configured.
```

Comment:
Instead of a List, could we do a Get using this expected name?

Or is there some more detailed implication where if any imageset exists, we need to always use imagesets? If so, I adjusted the comment a bit because that wasn't clear to me on first read
```suggestion
	// List the ImageSets to see if any exist. If at least one exists, then require
	// one which corresponds to the current version / variant. This ensures that
	// a user upgrading between versions of Operator won't accidentally move off
	// of ImageSets if they accidentally didn't create an ImageSet with the new expected name.
```

Reply:
yeah the list is so that we know if any ImageSets exist. I figured it might as well use what is returned in the list instead of doing a separate Get. I'll include the comment update